### PR TITLE
Resolves issue 1147

### PR DESCRIPTION
**Pull request by ryan-lynch**
Removed a hack that checks for @learningpool.com e-mail address to give permissions to resetPassword and replaced it with a call to hasPermission. Also removed resetPassword endpoint from the ignoreRoutes array.

### CODE REVIEW

File: lib/usermanager.js [hunk 1]
```diff
@@ -751,17 +766,20 @@ exports = module.exports = {
         return res.json({success: false, message: 'request body was not a valid object'});
       }
 
-      // TODO - Verify the current user has rights
-      // HACK -- Allowing @learningpool.com email addresses for now
-      if (user.email.indexOf('@learningpool.com') > -1) {
+      self.hasPermission('update', user, delta, function(err, isAllowed) {
+        if (err) {
```
Comment (brian-learningpool) [inline lib/usermanager.js]: What if `isAllowed` is `false`?

File: lib/usermanager.js [hunk 2]
```diff
@@ -53,6 +52,22 @@ exports = module.exports = {
...
+   * @param {objectid} userId
+   * @param {objectid} tenantId
+   * @param {object} a content item
+   * @param {callback} next (function (err, isAllowed))
+   */
+
+  hasPermission: function(action, user, contentItem, next) {
+    permissions = require('./permissions');
+    var resource = permissions.buildResourceString(user.tenant._id, '/api/user/resetpassword/' + contentItem);
+    permissions.hasPermission(user._id, action, resource, next);
```
Comment (brian-learningpool) [inline lib/usermanager.js]: I don't think the `contentItem` parameter is correct here if you're only looking to restrict access to the `/api/user/resetpassword` endpoint.

File: lib/usermanager.js [hunk 3]
```diff
@@ -53,6 +52,22 @@ exports = module.exports = {
...
+   *
+   * @param {string} action
+   * @param {objectid} userId
+   * @param {objectid} tenantId
+   * @param {object} a content item
+   * @param {callback} next (function (err, isAllowed))
+   */
+
+  hasPermission: function(action, user, contentItem, next) {
+    permissions = require('./permissions');
```
Comment (brian-learningpool) [inline lib/usermanager.js]: Hmmm.

Comment (ryan-lynch) [inline lib/usermanager.js]: Without a rather large refactor, I think this is the only way to resolve the problem right now.

File: lib/usermanager.js (hunk 2 above)
Comment (ryan-lynch) [inline lib/usermanager.js]: That's true, I don't think contentItem is necessary here.

File: lib/usermanager.js (hunk 1 above)
Comment (ryan-lynch) [inline lib/usermanager.js]: Woops, I'll handle this

File: lib/usermanager.js [hunk 4]
```diff
@@ -53,6 +52,22 @@ exports = module.exports = {
     'DuplicateUserError': DuplicateUserError
   },
 
+  /**
+   * checks if a user is permitted the action on the resource
+   *
+   * @param {string} action
```
Comment (ryangadams) [inline lib/usermanager.js]: this jsdoc needs updated to match the function details

File: lib/usermanager.js [hunk 5]
```diff
@@ -53,6 +52,20 @@ exports = module.exports = {
...
+   * @param {object} user
+   * @param {callback} next (function (err, isAllowed))
+   */
+
+  hasPermission: function(action, user, next) {
+    permissions = require('./permissions');
+    var resource = permissions.buildResourceString(user.tenant._id, '/api/user/resetpassword');
```
Comment (ryangadams) [inline lib/usermanager.js]: Now that I look at this shouldn't this resource be injected into hasPermission? 

Something like 
```
hasPermission: function(action, user, resource, next) {
  permissions = require('./permissions');
  var resource = permissions.buildResourceString(user.tenant._id, resource);
// etc.
```
Then on line 767
```
self.hasPermission('update', user, '/api/user/resetpassword', function(err, isAllowed) {
   // do stuff
}
```

File: lib/usermanager.js [hunk 6]
```diff
@@ -53,6 +52,21 @@ exports = module.exports = {
...
+   * checks if a user is permitted the action on the resource
+   *
+   * @param {string} action
+   * @param {object} user
+   * @param {string} resource
+   * @param {callback} next (function (err, isAllowed))
+   */
+
+  hasPermission: function(action, user, resource, next) {
+    permissions = require('./permissions');
```
Comment (brian-learningpool) [inline lib/usermanager.js]: You shouldn't do this:
https://github.com/alanjames1987/Node.js-Best-Practices#never-require-modules-inside-of-functions

Comment (brian-learningpool) [inline lib/usermanager.js]: Remove the circular dependency by removing the `require('./usermanager') and replacing calls to `usermanager` with `app.usermanager`.

**Comment by brian-learningpool** [comment]
+1

**Comment by ryangadams** [comment]
+1

**Pull request merged**